rbs/TypeTranslator.cc

```diff
@@ -0,0 +1,418 @@
+#include "TypeTranslator.h"
+#include "ast/Helpers.h"
+#include "core/GlobalState.h"
+#include "core/errors/internal.h"
+#include "core/errors/rewriter.h"
+
+using namespace sorbet::ast;
+
+namespace sorbet::rbs {
+
+namespace {
+
+bool hasTypeParam(core::MutableContext ctx, std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                  std::string_view name) {
+    return absl::c_any_of(typeParams, [&](const auto &param) { return param.second.toString(ctx.state) == name; });
+}
+
+sorbet::ast::ExpressionPtr typeNameType(core::MutableContext ctx,
+                                        std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                                        rbs_typename_t *typeName, bool isGeneric, core::LocOffsets loc) {
+    rbs_node_list *typePath = typeName->rbs_namespace->path;
+
+    ast::ExpressionPtr parent;
+
+    if (typeName->rbs_namespace->absolute) {
+        parent = ast::MK::Constant(loc, core::Symbols::root());
+    } else {
+        parent = ast::MK::EmptyTree();
+    }
+
+    if (typePath != nullptr) {
+        for (rbs_node_list_node *list_node = typePath->head; list_node != nullptr; list_node = list_node->next) {
+            rbs_node_t *node = list_node->node;
+
+            ENFORCE(node->type == RBS_AST_SYMBOL, "Unexpected node type `{}` in type name, expected `{}`",
+                    rbs_node_type_name(node), "Symbol");
+
+            rbs_ast_symbol_t *symbol = (rbs_ast_symbol_t *)node;
+            rbs_constant_t *name = rbs_constant_pool_id_to_constant(fake_constant_pool, symbol->constant_id);
+            std::string pathNameStr(name->start, name->length);
+            auto pathNameConst = ctx.state.enterNameConstant(pathNameStr);
+            parent = ast::MK::UnresolvedConstant(loc, std::move(parent), pathNameConst);
+        }
+    }
+
+    rbs_constant_t *name = rbs_constant_pool_id_to_constant(fake_constant_pool, typeName->name->constant_id);
+    std::string_view nameStr(name->start, name->length);
+
+    if (typePath == nullptr || typePath->length == 0) {
+        if (nameStr == "Array" && isGeneric) {
+            return ast::MK::T_Array(loc);
+        } else if (nameStr == "Hash" && isGeneric) {
+            return ast::MK::T_Hash(loc);
+        } else if (nameStr == "Set" && isGeneric) {
+            return ast::MK::T_Set(loc);
+        } else if (nameStr == "Class" && isGeneric) {
+            return ast::MK::T_Class(loc);
```

Comment:
Why are these the only `T::` types?

There are many more, listed in in `ClassOrModuleRef::isBuiltinGenericForwarder`.

Reply:
Added the missing types 👍